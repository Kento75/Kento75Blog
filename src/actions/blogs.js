import database from '../firebase/firebase';

// ADD_BLOG
export const ADD_BLOG = 'ADD_BLOG';

export const addBlog = (blog) => ({
	type: ADD_BLOG,
	blog
});

export const startAddBlog = (blogData = {}) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;
		const {
			title = '',
			body = '',
			createdAt = 0
		} = blogData;

		const blog = { title, body, createdAt };

		return database.ref(`users/${uid}/blogs`).push(blog).then((ref) => {
			dispatch(addBlog({ id: ref.key, ...blog }));
		});
	};
};

// REMOVE_BLOG
export const REMOVE_BLOG = 'REMOVE_BLOG';

export const removeBlog = ({ id } = {}) => ({
	type: REMOVE_BLOG,
	id
});

export const startRemoveBlog = ({ id }) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;

		return database.ref(`users/${uid}/blogs/${id}`).remove().then(() => {
			dispatch(removeBlog({ id }));
		});
	};
};

// EDIT_BLOG
export const EDIT_BLOG = 'EDIT_BLOG';

export const editBlog = (id, updates) => ({
	type: EDIT_BLOG,
	id,
	updates
});

export const startEditBlog = (id, updates) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;

		return database.ref(`users/${uid}/blogs/${id}`).update(updates).then(() => {
			dispatch(editBlog(id, updates));
		});
	};
};

// SET_BLOG
export const SET_BLOG = 'SET_BLOG';

export const setBlogs = (blogs) => ({
	type: SET_BLOG,
	blogs
});

export const startSetBlogs = () => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;

		return database.ref(`users/${uid}/blogs`).once('value').then((snapshot) => {
			const blogs = [];
			snapshot.forEach((blog) => {
				blogs.push({
					id: blog.key,
					...blog.val()
				});
			});
			dispatch(setBlogs(blogs));
		});
	};
};

// GET_BLOG - To fetch the blog based on id
export const startGetBlog = ({ id }) => {
	return () => {
		return database.ref(`users`).once('value').then((snapshot) => {
			let blog;
			snapshot.forEach((child) => {
				blog = child.val().blogs[id];
			});
			return blog;
		});
	};
};