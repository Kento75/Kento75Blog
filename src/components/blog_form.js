import React, { Component } from 'react';
import moment from 'moment';
import ReactQuill from 'react-quill';
import 'react-quill/dist/quill.snow.css';

export class BlogForm extends Component {
	constructor(props) {
		super(props);

		this.state = {
			title: props.blog ? props.blog.title : '',
			body: props.blog ? props.blog.body : '',
			createdAt: props.blog ? moment(props.blog.createdAt) : moment(),
			error: ''
		};

		this.handleFormSubmit = this.handleFormSubmit.bind(this);
		this.handleTitleChange = this.handleTitleChange.bind(this);
		this.handleBodyChange = this.handleBodyChange.bind(this);
	}

	handleFormSubmit(e) {
		e.preventDefault();
		if (this.state.title && this.state.body) {
			this.props.onSubmit({
				title: this.state.title,
				body: this.state.body,
				createdAt: this.state.createdAt.valueOf()
			});
			this.setState({ error: '' });
		} else {
			this.setState({ error: 'Please provide title and body!' });
		}
	}

	handleTitleChange(e) {
		this.setState({ title: e.target.value });
	}

	handleBodyChange(value) {
		this.setState({ body: value });
	}

	render() {
		return (
			<form className="form" onSubmit={this.handleFormSubmit}>
				{this.state.error && <p className="form__error">{this.state.error}</p>}
				<input className="text-input" type="text" placeholder="Post Title" value={this.state.title} onChange={this.handleTitleChange} autoFocus />
				<ReactQuill value={this.state.body} onChange={this.handleBodyChange} />
				<input className="button" type="submit" value="Save Post" />
			</form>
		);
	}
}

export default BlogForm;