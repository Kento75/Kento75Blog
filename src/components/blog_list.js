import React, { Component } from 'react';
import { connect } from 'react-redux';

import BlogListItem from './blog_list_item';
import selectBlogs from '../selectors/blogs';

export const BlogList = (props) => (
	<div className="content-container">
		<div className="list-header">
			<div className="show-for-mobile">Blogs</div>
			<div className="show-for-desktop">Blog</div>
			<div className="show-for-desktop">Posted on</div>
		</div>
		<div className="list-body">
			{
				props.blogs.length === 0 ? (
					<p className="list-item list-item--message">No Blogs. Write one now!</p>
				) : (
					props.blogs.map((blog) => {
						return <BlogListItem key={blog.id} {...blog} />;
					})
				)
			}
		</div>
	</div>
);

const mapStateToProps = (state) => ({
	blogs: selectBlogs(state.blogs, state.filters)
});

export default connect(mapStateToProps)(BlogList);