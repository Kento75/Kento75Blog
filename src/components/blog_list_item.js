import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';

export const BlogListItem = ({ id, title, createdAt }) => (
	<Link className="list-item" to={`/edit/${id}`}>
		<div>
			<h4 className="list-item__title">{title}</h4>
			<span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
		</div>
	</Link>
);

const mapStateToProps = (state) => ({
	blogs: state.blogs
});

export default connect(mapStateToProps)(BlogListItem);