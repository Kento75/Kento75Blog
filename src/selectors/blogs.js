export default (blogs, { text, sortBy }) => {
	return blogs.filter((blog) => {
		const titleMatch = blog.title.toLowerCase().includes(text.toLowerCase());

		return titleMatch;
	}).sort((a, b) => {
		if (sortBy === 'date') {
			return a.createdAt < b.createdAt ? 1 : -1;
		}
	});
};