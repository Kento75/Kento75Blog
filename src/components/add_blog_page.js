import React, { Component } from 'react';
import { connect } from 'react-redux';

import BlogForm from './blog_form';
import { startAddBlog } from '../actions/blogs';

export class AddBlogPage extends Component {
	constructor(props) {
		super(props);

		this.handleSubmit = this.handleSubmit.bind(this);
	}

	handleSubmit(blog) {
		this.props.startAddBlog(blog);
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Add Blog</h1>
					</div>
				</div>
				<div className="content-container">
					<BlogForm onSubmit={this.handleSubmit} />
				</div>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({
	startAddBlog: (blog) => dispatch(startAddBlog(blog))
});

export default connect(undefined, mapDispatchToProps)(AddBlogPage);