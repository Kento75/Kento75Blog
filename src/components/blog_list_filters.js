import React, { Component } from 'react';
import { connect } from 'react-redux';

import { setTextFilter, sortByDate, sortByTitle } from '../actions/filters';

export class BlogListFilters extends Component {
	constructor(props) {
		super(props);

		this.handleSortByChange = this.handleSortByChange.bind(this);
		this.handleTextFilterChange = this.handleTextFilterChange.bind(this);
	}

	handleSortByChange(e) {
		switch (e.target.value) {
		case 'date':
			this.props.sortByDate();
			break;
		case 'title':
			this.props.sortByTitle();
			break;
		default:
			break;
		}	
	}

	handleTextFilterChange(e) {
		this.props.setTextFilter(e.target.value);
	}

	render() {
		return (
			<div className="content-container">
				<div className="input-group">
					<div className="input-group__item">
						<input className="text-input" type="text" placeholder="Search Posts..." value={this.props.filters.text} onChange={this.handleTextFilterChange} />
					</div>
					<div className="input-group__item">
						<select className="select" value={this.props.filters.sortBy} onChange={this.handleSortByChange}>
							<option value="date">Date</option>
							<option value="amount">Title</option>
						</select>
					</div>
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({
	filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByTitle: () => dispatch(sortByTitle())
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogListFilters);