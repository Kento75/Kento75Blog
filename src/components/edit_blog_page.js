import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import BlogForm from './blog_form';
import { startEditBlog, startRemoveBlog } from '../actions/blogs';

export class EditBlogPage extends Component {
	constructor(props) {
		super(props);

		this.handleSubmit = this.handleSubmit.bind(this);
		this.handleClick = this.handleClick.bind(this);
	}

	handleSubmit(blog) {
		this.props.startEditBlog(this.props.blog.id, blog);
		this.props.history.push('/');
	}	

	handleClick(e) {
		this.props.startRemoveBlog({ id: this.props.blog.id });
		this.props.history.push('/');
	}
	
	render() {
		const readableLink = `${window.location.origin}/read/${this.props.blog.id}`;
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Edit Blog</h1>
						Blog readable at <Link to={`/read/${this.props.blog.id}`}>{readableLink}</Link>
					</div>
				</div>
				<div className="content-container">
					<BlogForm blog={this.props.blog} onSubmit={this.handleSubmit} />
					<button className="button button--secondary" onClick={this.handleClick}>Remove Blog</button>
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => ({
	blog: state.blogs.find((blog) => blog.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
	startEditBlog: (id, updates) => dispatch(startEditBlog(id, updates)),
	startRemoveBlog: (id) => dispatch(startRemoveBlog(id)) 
});

export default connect(mapStateToProps, mapDispatchToProps)(EditBlogPage);