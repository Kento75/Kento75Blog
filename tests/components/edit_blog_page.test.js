import React from 'react';
import { shallow } from 'enzyme';

import { EditBlogPage } from '../../src/components/edit_blog_page';
import blogs from '../fixtures/blogs';

let wrapper, history, startEditBlog, startRemoveBlog;

beforeEach(() => {
	history = { push: jest.fn() };
	startEditBlog = jest.fn();
	startRemoveBlog = jest.fn();
	wrapper = shallow(<EditBlogPage 
		blog={blogs[1]}
		history={history} 
		startEditBlog={startEditBlog} 
		startRemoveBlog={startRemoveBlog} 
	/>);
});

test('should render EditBlogPage correctly', () => {
	expect(wrapper).toMatchSnapshot();
});

test('should handle editBlog', () => {
	wrapper.find('BlogForm').prop('onSubmit')(blogs[1]);
	expect(history.push).toHaveBeenLastCalledWith('/');
	expect(startEditBlog).toHaveBeenLastCalledWith(blogs[1].id, blogs[1]);
});

test('should handle removeBlog', () => {
	wrapper.find('button').simulate('click');
	expect(history.push).toHaveBeenLastCalledWith('/');
	expect(startRemoveBlog).toHaveBeenLastCalledWith({ id: blogs[1].id });
});