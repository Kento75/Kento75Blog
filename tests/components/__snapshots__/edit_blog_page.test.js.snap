// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditBlogPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Blog
      </h1>
      Blog readable at 
      <Link
        replace={false}
        to="/read/pqr567"
      >
        null/read/pqr567
      </Link>
    </div>
  </div>
  <div
    className="content-container"
  >
    <BlogForm
      blog={
        Object {
          "body": "Directives, Lazy Load",
          "createdAt": 345600000,
          "id": "pqr567",
          "title": "Write a Angular app",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Blog
    </button>
  </div>
</div>
`;
